tools/pythonpkg/src/native/python_conversion.cpp

```diff
@@ -523,113 +478,587 @@ Value TransformPythonValue(py::handle ele, const LogicalType &target_type, bool
+			return;
+		}
+		Value result_val;
+		PythonValueConversion::HandleString(result_val, result_type, value);
+		FallbackValueConversion(result, result_offset, std::move(result_val));
+	}
+
+	static void HandleDate(Vector &result, const idx_t &result_offset, PyDate &date) {
+		auto &result_type = result.GetType();
+		switch (result_type.id()) {
+		case LogicalTypeId::DATE:
+			FlatVector::GetData<date_t>(result)[result_offset] = date.ToDate();
+			break;
+		default: {
+			auto value = date.ToDuckValue();
+			result.SetValue(result_offset, value);
+			break;
+		}
+		}
+	}
+
+	static void HandleTime(Vector &result, const idx_t &result_offset, PyTime &time) {
+		auto &result_type = result.GetType();
+		switch (result_type.id()) {
+		case LogicalTypeId::TIME:
+			FlatVector::GetData<dtime_t>(result)[result_offset] = time.ToDuckTime();
+			break;
+		default: {
+			auto value = time.ToDuckValue();
+			result.SetValue(result_offset, value);
+			break;
+		}
+		}
+	}
+
+	static void HandleBlob(Vector &result, const idx_t &result_offset, const_data_ptr_t blob, idx_t blob_size) {
+		auto &result_type = result.GetType();
+		switch (result_type.id()) {
+		case LogicalTypeId::BLOB:
+			FlatVector::GetData<string_t>(result)[result_offset] =
+			    StringVector::AddStringOrBlob(result, const_char_ptr_cast(blob), blob_size);
+			break;
+		default: {
+			auto value = Value::BLOB(blob, blob_size);
+			result.SetValue(result_offset, value);
+			break;
+		}
+		}
+	}
+
+	static void HandleDateTime(Vector &result, const idx_t &result_offset, PyDateTime &datetime) {
+		auto &result_type = result.GetType();
+		switch (result_type.id()) {
+		case LogicalTypeId::TIMESTAMP:
+			FlatVector::GetData<timestamp_t>(result)[result_offset] = datetime.ToTimestamp();
+			break;
+		case LogicalTypeId::TIME:
+			FlatVector::GetData<dtime_t>(result)[result_offset] = datetime.ToDuckTime();
+			break;
+		case LogicalTypeId::DATE:
+			FlatVector::GetData<date_t>(result)[result_offset] = datetime.ToDate();
+			break;
+		default: {
+			auto value = datetime.ToDuckValue(result_type);
+			result.SetValue(result_offset, value);
+			break;
+		}
+		}
+	}
+
+	template <bool IS_LIST>
+	static void HandleListFast(Vector &result, const idx_t &result_offset, py::handle ele, idx_t list_size) {
+		auto &result_type = result.GetType();
+		if (result_type.id() == LogicalTypeId::ARRAY) {
+			idx_t array_size = ArrayType::GetSize(result_type);
+			if (list_size != array_size) {
+				throw InvalidInputException("Python Conversion Failure: Array size mismatch - expected an array of "
+				                            "size %d, but got a list of size %d",
+				                            array_size, list_size);
+			}
+			auto &child_array = ArrayVector::GetEntry(result);
+			idx_t start_offset = result_offset * array_size;
+			for (idx_t i = 0; i < list_size; i++) {
+				auto child_ele = IS_LIST ? PyList_GetItem(ele.ptr(), i) : PyTuple_GetItem(ele.ptr(), i);
```

Comment:
I'm not sure if we should do this, part of Pythons strength is ducktyping, I believe we also return the LIST PythonObjectType for duck-typed lists.
When we use this method on something that is not `list` we'll likely crash

Reply:
But perhaps we can do a fast-er path for this, after verifying that is actually an instance of  `py::list`

Reply:
That should already happen in determining the `PythonObjectType`:

```cpp
	} else if (py::isinstance<py::list>(ele)) {
		return PythonObjectType::List;
	} else if (py::isinstance<py::tuple>(ele)) {
		return PythonObjectType::Tuple;
```

But we could add another check + internal exception here, sure

Reply:
You're right, I thought we used "is list-like" here